Mathlib/AlgebraicGeometry/PrimeSpectrum/TensorProduct.lean

```diff
@@ -0,0 +1,80 @@
+Authors: Andrew Yang
+-/
+import Mathlib.RingTheory.SurjectiveOnStalks
+import Mathlib.AlgebraicGeometry.PrimeSpectrum.Basic
+
+/-!
+
+# Lemmas regarding the prime spectrum of tensor products
+
+## Main result
+- `PrimeSpectrum.embedding_tensorProductTo_of_surjectiveOnStalks`:
+  If `R →+* T` is surjective on stalks,
+  then `Spec(S ⊗[R] T) → Spec S × Spec T` is a topological embedding
+  (where `Spec S × Spec T` is the cartesian product with the product topology).
+-/
+
+variable (R S T) [CommRing R] [CommRing S] [Algebra R S]
+variable [CommRing T] [Algebra R T] [Algebra S T] [IsScalarTower R S T]
+
+open TensorProduct
+
+/-- The canonical map from `Spec(S ⊗[R] T)` to the cartesian product `Spec S × Spec T`. -/
+noncomputable
+def PrimeSpectrum.tensorProductTo (x : PrimeSpectrum (S ⊗[R] T)) :
+    PrimeSpectrum S × PrimeSpectrum T :=
+  ⟨comap (algebraMap _ _) x, comap Algebra.TensorProduct.includeRight.toRingHom x⟩
+
+lemma PrimeSpectrum.continuous_tensorProductTo : Continuous (tensorProductTo R S T) :=
+  (comap _).2.prod_mk (comap _).2
+
+variable (hRT : (algebraMap R T).SurjectiveOnStalks)
+
+lemma PrimeSpectrum.embedding_tensorProductTo_of_surjectiveOnStalks_aux
+    (p₁ p₂ : PrimeSpectrum (S ⊗[R] T))
+    (h : tensorProductTo R S T p₁ = tensorProductTo R S T p₂) :
+    p₁ ≤ p₂ := by
+  let g : T →+* S ⊗[R] T := Algebra.TensorProduct.includeRight.toRingHom
+  intros x hxp₁
+  by_contra hxp₂
+  obtain ⟨t, r, a, ht, e⟩ := RingHom.exists_mul_eq_tmul_of_surjectiveOnStalks hRT x
+    (p₂.asIdeal.comap g) inferInstance
+  have h₁ : a ⊗ₜ[R] t ∈ p₁.asIdeal := e ▸ p₁.asIdeal.mul_mem_left (1 ⊗ₜ[R] (r • t)) hxp₁
+  have h₂ : a ⊗ₜ[R] t ∉ p₂.asIdeal := e ▸ p₂.asIdeal.primeCompl.mul_mem ht hxp₂
+  rw [← mul_one a, ← one_mul t, ← Algebra.TensorProduct.tmul_mul_tmul] at h₁ h₂
+  have h₃ : t ∉ p₂.asIdeal.comap g := fun h ↦ h₂ (Ideal.mul_mem_left _ _ h)
+  have h₄ : a ∉ p₂.asIdeal.comap (algebraMap S (S ⊗[R] T)) :=
+    fun h ↦ h₂ (Ideal.mul_mem_right _ _ h)
+  replace h₃ : t ∉ p₁.asIdeal.comap g := by
+    rwa [show p₁.asIdeal.comap g = p₂.asIdeal.comap g from congr($h.2.1)]
+  replace h₄ : a ∉ p₁.asIdeal.comap (algebraMap S (S ⊗[R] T)) := by
+    rwa [show p₁.asIdeal.comap (algebraMap S (S ⊗[R] T)) = p₂.asIdeal.comap _ from congr($h.1.1)]
+  exact p₁.asIdeal.primeCompl.mul_mem h₄ h₃ h₁
+
+lemma PrimeSpectrum.embedding_tensorProductTo_of_surjectiveOnStalks :
+    Embedding (tensorProductTo R S T) := by
+  refine ⟨?_, fun p₁ p₂ e ↦
+    (embedding_tensorProductTo_of_surjectiveOnStalks_aux R S T hRT p₁ p₂ e).antisymm
+      (embedding_tensorProductTo_of_surjectiveOnStalks_aux R S T hRT p₂ p₁ e.symm)⟩
+  let g : T →+* S ⊗[R] T := Algebra.TensorProduct.includeRight.toRingHom
+  refine ⟨(continuous_tensorProductTo R S T).le_induced.antisymm ?_⟩
+  conv_rhs => rw [isBasis_basic_opens.eq_generateFrom]
+  rw [TopologicalSpace.le_generateFrom_iff_subset_isOpen]
+  rintro _ ⟨_, ⟨f, rfl⟩, rfl⟩
+  rw [Set.mem_setOf_eq, @isOpen_iff_forall_mem_open]
```

Comment:
Could this part be extracted to a lemma about when a topology `t` satisfies `t \le zariskiTopology`?

Reply:
Yes, thanks